applications/OptimizationApplication/python_scripts/algorithms/algorithm_gradient_projection.py

```diff
@@ -1,120 +1,228 @@
+        self.correction_size = settings["correction_size"].GetDouble()
+
+    def GetMinimumBufferSize(self) -> int:
+        return 2
+
+    def Check(self):
+        pass
+    
+    @time_decorator()
+    def Initialize(self):
+        self.converged = False
+        self.__obj_val = None
+        self.__objective.Initialize()
+        self.__objective.Check()
+        CallOnAll(self.__constraints_list, StandardizedConstraint.Initialize)
+        CallOnAll(self.__constraints_list, StandardizedConstraint.Check)
+        self.master_control.Initialize()
+        self.__control_field = self.master_control.GetControlField()
+        self.algorithm_data = ComponentDataView("algorithm", self._optimization_problem)
+
+    @time_decorator()
+    def Finalize(self):
+        pass
+    
+    @time_decorator()
+    def ComputeSearchDirection(self, obj_grad: KratosOA.CollectiveExpression, constr_grad: 'list[KratosOA.CollectiveExpression]') -> KratosOA.CollectiveExpression:
+        active_constraints_list = [self.__constraints_list[i] for i in range(len(self.__constraints_list)) if self.__constr_value[i] >= 0.0]
+        number_of_active_constraints = len(active_constraints_list)
+        if not number_of_active_constraints:
+            search_direction = obj_grad.Clone() * -1.0
+            correction = obj_grad.Clone() * 0.0
+        else:
+            constraint_violations = Kratos.Vector(number_of_active_constraints)
+            for i, active_constraint in enumerate(active_constraints_list):
+                    constraint_violations[i] = active_constraint.GetScaledViolationValue()
+
+            # compute the projected search direction and correction
+            ntn = Kratos.Matrix(number_of_active_constraints, number_of_active_constraints)
+            for i in range(number_of_active_constraints):
+                for j in range(i, number_of_active_constraints):
+                    ntn[i, j] = KratosOA.ExpressionUtils.InnerProduct(constr_grad[i], constr_grad[j])
+                    ntn[j, i] = ntn[i, j]
+
+                # get the inverse of ntn
+                ntn_inverse = Kratos.Matrix(number_of_active_constraints, number_of_active_constraints)
+
+                # create the identity matrix
+                identity_matrix = Kratos.Matrix(number_of_active_constraints, number_of_active_constraints, 0.0)
+                for i in range(number_of_active_constraints):
+                    identity_matrix[i, i] = 1.0
+
+                # solve for inverse of ntn
+                self.linear_solver.Solve(ntn, ntn_inverse, identity_matrix)
+
+                search_direction = - (obj_grad - self.__CollectiveListVectorProduct(constr_grad, ntn_inverse * self.__CollectiveListCollectiveProduct(constr_grad, obj_grad)))
+                correction = - self.__CollectiveListVectorProduct(constr_grad, ntn_inverse * constraint_violations)
+        correction_norm = KratosOA.ExpressionUtils.NormInf(correction)
```

Comment:
Should not it be l2_norm ?

Reply:
Inf_norm is a more strict norm compare to l2_norm. Let's say I have vector [0,0, ... , 100, 0, 0, n] and the size is 10e5. So L2 norm would be 100 / 10e5 and it is going to be small. Still computed correction for the parameter can ve very large alpha * 100. In case of inf_norm, it makes sure that all parameters change is under certain limit. Still, both can be used and it is always a discussion which one is better.